### PR TITLE
[Feature](bangc-ops): add new op diff_iou_rotated_sort_vertices_forward

### PR DESCRIPTION
**Pull request by WsqRichards1**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

add new op diff_iou_rotated_sort_vertices_forward

## 2. Modification

1. new file:   bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward.cpp
2. new file:   bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward.h
3. new file:   bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward_block.mlu
4. modified:   bangc-ops/mlu_op.h
5. new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward.cpp
6. new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward.h
7. new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/diff_iou_rotated_sort_vertices_forward/test_case/case_0.prototxt
8. new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/diff_iou_rotated_sort_vertices_forward/test_case/case_1.prototxt
9. new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/diff_iou_rotated_sort_vertices_forward/test_case/case_2.prototxt
10. new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/diff_iou_rotated_sort_vertices_forward/test_case/case_3.prototxt
11. new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/diff_iou_rotated_sort_vertices_forward/test_case/case_4.prototxt
12. new file:   docs/bangc-docs/design_docs/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward.md
13. new file:   docs/bangc-docs/design_docs/diff_iou_rotated_sort_vertices_forward/res/nram_space.png
14. new file:   docs/bangc-docs/design_docs/diff_iou_rotated_sort_vertices_forward/res/pingpong.png
15. new file:   docs/bangc-docs/design_docs/diff_iou_rotated_sort_vertices_forward/res/sort_vertices.png
16. new file:   docs/bangc-docs/design_docs/diff_iou_rotated_sort_vertices_forward/res/sort_vertices1.png
17. modified:   docs/bangc-docs/user_guide/9_operators/index.rst
18. new file:   docs/bangc-docs/user_guide/images/sort_vertices.png

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [x] diff1: diff1 <= 3e-3
- [x] diff2: diff2 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |          BLOCK       |
|        3       |Layouts                               |          ARRAY     |
|        4       |Whether multi-dimensions are supported|                    no                   |
|        5       |Whether element zero is supported     |                  N=0 返回 MLUOP_STATUS_SUCCESS，其余情况直接返回MLUOP_STATUS_BAD_PARAM                     |
|        6       |Data type(half/float)                 |           float           |
|        7       |Whether there is size limit           | vertices_desc.dims[0] == mask_desc.dims[0] == num_valid_desc.dims[0] == idx_desc.dims[0]<br>vertices_desc.dims[1] == mask_desc.dims[1] == num_valid_desc.dims[1] == idx_desc.dims[1]<br>    vertices_desc.dims[2] == mask_desc.dims[2] == 24<br> vertices_desc.dims[3] == 2 <br> idx_desc.dims[2] == 9                               |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |            通过                 |
| Whether illegal parameters are passed           |     Normal error    |                   通过          |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |      float           |    通过      |    [----------] Global test environment tear-down<br>[ SUMMARY  ] Total 252 cases of 1 op(s).<br>ALL PASSED.<br>[==========] 445 test cases from 1 test suite ran. (6732 ms total)<br>[  PASSED  ] 445 test cases.      |
|Zero element test    |Whether to support this test      |     通过     |   vertices_desc.dims[1] = 0 [2023-5-11 15:52:26] [MLUOP] [Warning]:The result error is empty, it means output shape is 0 in pb, and skip compute result error. <br> vertices_desc.dims[0] = 0, vertices_desc.dims[2] = 0 [2023-5-11 15:51:43] [MLUOP] [Error]:"MLUOP_STATUS_BAD_PARAM in mluOpDiffIouRotatedSortVerticesForward    |
|Stability test       |--gtest_repeat=NUM<br>--thread=NUM|     通过     |   ./mluop_gtest --gtest_filter=\*diff_iou_rotated_sort_vertices_forward* --cases_dir=../../pb/ --gtest_repeat=100 --thread=100<br>   [----------] Global test environment tear-down<br>[ SUMMARY  ] Total 44500 cases of 100 op(s).<br>ALL PASSED.<br>[==========] 1 test case from 1 test suite ran. (1408 ms total)<br>[  PASSED  ] 1 test case.     |
|Mult-platform test   |MLU370/MLU590                     |     通过     |  MLU370：<br>    [----------] Global test environment tear-down<br>[ SUMMARY  ] Total 445 cases of 1 op(s).<br>ALL PASSED.<br>[==========] 445 test cases from 1 test suite ran. (6732 ms total)<br>[  PASSED  ] 445 test cases. <br>  MLU590:<br> [----------] Global test environment tear-down<br>[ SUMMARY  ] Total 445 cases of 1 op(s).<br>ALL PASSED.<br>[==========] 252 test cases from 1 test suite ran. (6094 ms total)<br>[  PASSED  ] 445 test cases. |
|Nan/INF test         |Whether to support this test      |     通过     |          |
|Memory leak check    |Test result                       |     通过     |          |
|Code coverage check  |Test result                       |     通过     | Filename [Sort by name]               Line Coverage [Sort by_line_coverage]  Functions [Sort by_function_coverage<br>diff_iou_rotated_sort_vertices_forward_block.mlu[98.9%]  98.9 % / 357   100.0 %/ 9         |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Data_type|Shape|
|-------|----|----|-----|----|----|-----|
|diff_iou_sort_vertices_forward|  23  |  44.24  |  0.00126   |  0.0002169  |   float  |vertices:[1, 35,24, 2]<br>mask:[1, 35, 24] <br>valid_num: [1, 35] <br> idx:[1, 35, 9] |
|diff_iou_sort_vertices_forward|  77 |  37.41 |  0.0024   | 0.00041  |  float  |vertices:[1, 224,24, 2]<br>mask:[1, 224, 24] <br>valid_num: [1, 224] <br> idx:[1, 224, 9] |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Data_type|Shape|
|-------|----|----|----|----|----|-----|
|diff_iou_sort_vertices_forward|  14  |  18.788  |  0.000231   | 0.000148   |  float  |vertices:[1, 35,24, 2]<br>mask:[1, 35, 24] <br>valid_num: [1, 35] <br> idx:[1, 35, 9] |
|diff_iou_sort_vertices_forward|  39  |  17.793  |  0.000531   | 0.000341  |  float  |vertices:[1, 224,24, 2]<br>mask:[1, 224, 24] <br>valid_num: [1, 224] <br> idx:[1, 224, 9] |

### 3.4 Summary Analysis
所有测试已通过。

### CODE REVIEW

File: bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward_block.mlu [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+  // |           nram_vertices              |            nram_mask           |
+  // |--------------------------------------|--------------------------------|
+  // | deal_num * dim_m * 2 * sizeof(T)     | deal_num * dim_m * sizeof(bool)|
+  // |--------------------------------------|--------------------------------|
+  // |           nram_num_valid             |           nram_idx             |
+  // |        deal_num * sizeof(int)        |   9 * deal_num * sizeof(int)   |
+  // |--------------------------------------|--------------------------------|
+  int deal_num = MAX_NRAM_SIZE / 2 /
+                 (4 * dim_m * sizeof(T) + dim_m * sizeof(bool) +
+                  (1 + MAX_NUM_VERT_IDX) * sizeof(int));
```
Comment (Unireverse) [inline bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward_block.mlu:400]: 这里的空间划分好像和上面的图不太一致，我化简的结果似乎是 MAX_NRAM_SIZE / (80 + 2 * dim_m + 8 * dim_m * sizeof(T))，不知道是不是我的理解问题

File: bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward_block.mlu [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+#if (__BANG_ARCH__ != 372)
+    __bang_div((float *)nram_dst, (float *)nram_src0, (float *)nram_src1,
+               deal_num);
+#else
+    __bang_recip((float *)nram_dst, (float *)nram_src1, deal_num);
+    __bang_mul((float *)nram_dst, (float *)nram_src0, (float *)nram_dst,
+               deal_num);
+#endif
+  }
+}
```
Comment (Unireverse) [inline bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward_block.mlu:67]: 这个写法有点问题，half输入的时候会挂掉，建议把float改成T，然后在host把half防掉

Comment (DanieeelLiu) [reply bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward_block.mlu]: Fixed

File: docs/bangc-docs/design_docs/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward.md [hunk 3]
```diff
@@ -392,30 +392,30 @@ mluOpDiffIouRotatedSortVerticesForward(mluOpHandle_t handle,
...
   int deal_num_valid_num = deal_num;
-  int max_nram_size = MAX_NRAM_SIZE -  4 * deal_num * 24 * sizeof(T);
+  int max_nram_size = MAX_NRAM_SIZE -  3 * deal_num * 24 * sizeof(T);
   int pingpong_offset = max_nram_size / 2;
   
   T *nram_pub_space = (T *)nram_buffer;
   
   // ping/pong
-  char *nram_vertices = (char *)(nram_pub_space + 4 * deal_num * 24);
+  char *nram_vertices = (char *)(nram_pub_space + 3 * deal_num * 24);
```
Comment (Unireverse) [inline docs/bangc-docs/design_docs/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward.md:418]: 这里好像少乘了sizeof (T)

File: bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward_block.mlu (hunk 1 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward_block.mlu]: 已沟通说明，没问题

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -12518,6 +12518,85 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
                          const mluOpTensorDescriptor_t data_col_desc,
                          void *data_col);
 
+// Group:DiffIouRotatedSortVerticesForward
+/*!
+ * @brief Sort the effective vertices of the polygon formed by the intersection of two boxes,
```
[the remark this answers is not in the document]
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -12518,6 +12518,85 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ * and output the sorted vertex index.
```
[the remark this answers is not in the document]
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -12518,6 +12518,85 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in the
+ * diff_iou_rotated_sort_vertices_forward operation. For detailed information, see ::mluOpHandle_t.
+ *
+ * @param[in] vertices_desc
+ * The descriptor of input tensor \b vertices, which contains dimension, data type and data layout.
```
[the remark this answers is not in the document]
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -12518,6 +12518,85 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ *
+ * @param[in] vertices
+ * Pointer to the MLU memory that stores the \b vertices tensor.
```
[the remark this answers is not in the document]
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -12518,6 +12518,85 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ *
+ * @param[in] mask_desc
+ * The descriptor of input tensor \b mask, which contains dimension, data type and data layout.
```
[the remark this answers is not in the document]
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -12518,6 +12518,85 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ *
+ * @param[in] mask
+ * Pointer to the MLU memory that stores the \b mask tensor.
```
[the remark this answers is not in the document]
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -12518,6 +12518,85 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ *
+ * @param[in] num_valid_desc
+ * The descriptor of input tensor \b num_valid, which contains dimension, data type and data layout.
```
[the remark this answers is not in the document]
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -12518,6 +12518,85 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ *
+ * @param[in] num_valid
+ * Pointer to the MLU memory that stores the \b num_valid tensor.
+ *
+ * @param[out] idx_desc
+ * The descriptor of output tensor \b idx, which contains dimension, data type and data layout.
```
[the remark this answers is not in the document]
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -12518,6 +12518,85 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ *
+ * @param[out] idx
+ * Pointer to the MLU memory that stores the \b idx tensor.
```
[the remark this answers is not in the document]
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 13]
```diff
@@ -132,6 +132,12 @@ mluOpDiv
 - ``i`` 表示一个多维数组的索引，表示多维张量，例如在4维时可以表示（n,c,h,w）。
 - ``xi``、``yi``、``zi`` 表示多维数组中 ``i`` 索引处的元素。
 
+.. _diff_iou_rotated_sort_vertices_forward:
+
+mluOpDiffIouRotatedSortVerticesForward
+-----------------------------
+该算子为 `diff_iou_rotated` 整个计算过程中的一步，完成对两个box相交构成的多边形有效顶点排序（参考算法按逆时针排序所有有效顶点），输出排序后顶点索引。
```
[the remark this answers is not in the document]
Comment (DanieeelLiu) [reply docs/bangc-docs/user_guide/9_operators/index.rst]: Fixed

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 14]
```diff
@@ -132,6 +132,16 @@ mluOpDiv
...
 
+.. _diff_iou_rotated_sort_vertices_forward:
+
+mluOpDiffIouRotatedSortVerticesForward
+------------------------------------------
+该算子为 ``diff_iou_rotated`` 整个计算过程中的一步，完成对两个box相交构成的多边形有效顶点排序（参考算法按逆时针排序所有有效顶点），输出排序后顶点索引。
+示意图如下：
+
+.. figure:: ../images/sort_vertices.*
+   顶点逆时针排序示意图
```
Comment (AquamarineSu) [inline docs/bangc-docs/user_guide/9_operators/index.rst:143]:
```suggestion

   顶点逆时针排序示意图
```

**Review by AquamarineSu — APPROVED** [review]

File: bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward_block.mlu [hunk 15]
```diff
@@ -0,0 +1,396 @@
+/*************************************************************************
+ * Copyright (C) [2022] by Cambricon, Inc.
```
Comment (tudejiang79) [inline bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward_block.mlu:2]: 2023

File: bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward.h [hunk 16]
```diff
@@ -0,0 +1,53 @@
+/*************************************************************************
+ * Copyright (C) [2022] by Cambricon, Inc.
```
Comment (tudejiang79) [inline bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward.h:2]: 2023

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward.cpp [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+  interface_timer_.stop();
+  VLOG(4) << "[DiffIouRotatedSortVerticesForwardExecutor] call compute() end.";
+}
+
+/*
+compare normalized vertices (vertices around (0,0))
+if vertex1 < vertex2 return true.
+order: minimum at x-aixs, become larger in anti-clockwise direction
+*/
+bool DiffIouRotatedSortVerticesForwardExecutor
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward.cpp:64]: 建议都写成类内函数

File: bangc-ops/mlu_op.h [hunk 18]
```diff
@@ -12518,6 +12518,85 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ * Pointer to the MLU memory that stores the tensor \b num_valid.
+ *
+ * @param[out] idx_desc
+ * The descriptor of output tensor \b idx, which contains dimension, data type, and data layout.
+ *
+ * @param[out] idx
+ * Pointer to the MLU memory that stores the tensor \b idx.
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
```
Comment (PetrelYy) [inline bangc-ops/mlu_op.h:12556]: 漏了 MLUOP_STATUS_ARCH_MISMATCH

File: bangc-ops/mlu_op.h [hunk 19]
```diff
@@ -12518,6 +12518,85 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
 
+// Group:DiffIouRotatedSortVerticesForward
+/*!
+ * @brief Sorts the effective vertices of the polygon formed by the intersection of two boxes,
+ * and outputs the sorted vertex index.
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in the
+ * diff_iou_rotated_sort_vertices_forward operation. For detailed information, see ::mluOpHandle_t.
+ *
```
Comment (PetrelYy) [inline bangc-ops/mlu_op.h:12529]: 按已有格式，param handle~input/output 中间不空行

File: bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward_block.mlu (hunk 15 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward_block.mlu]: Fixed

File: bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward.h (hunk 16 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/diff_iou_rotated_sort_vertices_forward/diff_iou_rotated_sort_vertices_forward.h]: Fixed

File: bangc-ops/mlu_op.h (hunk 18 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h (hunk 19 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

**Review by tudejiang79 — APPROVED** [review]

**Review by PetrelYy — APPROVED** [review]

**Review by Unireverse — APPROVED** [review]

**Pull request merged**